electrum/plugins/satochip/qt.py

```diff
@@ -0,0 +1,1032 @@
+
+
+class WCSatochipSetup(WalletWizardComponent):
+    def __init__(self, parent, wizard):
+        WalletWizardComponent.__init__(
+            self, parent, wizard, title=_('Satochip Setup'))
+        _logger.info('[WCSatochipSetup] __init__()')  # debugsatochip
+
+        self.plugins = wizard.plugins
+        self.plugin = self.plugins.get_plugin('satochip')
+
+        self.layout().addWidget(WWLabel('Done'))
+
+        self._busy = True
+
+    def on_ready(self):
+        _logger.info('[WCSatochipSetup] on_ready()')  # debugsatochip
+        current_cosigner = self.wizard.current_cosigner(self.wizard_data)
+        settings = current_cosigner['satochip_setup_settings']
+        # method = current_cosigner['satochip_init']
+        _name, _info = current_cosigner['hardware_device']
+        device_id = _info.device.id_
+        # debugsatochip
+        _logger.info(f'[WCSatochipSetup] on_ready() device_id: {device_id}')
+
+        client = self.plugins.device_manager.client_by_id(
+            device_id, scan_now=False)
+        client.handler = self.plugin.create_handler(self.wizard)
+
+        def initialize_device_task(settings, device_id, client):
+            try:
+                # self.plugin._initialize_device(settings, method, device_id, handler)
+                self.plugin._setup_device(settings, device_id, client)
+                _logger.info(
+                    '[WCSatochipSetup] initialize_device_task() Done initialize device')
+                self.valid = True
+                self.wizard.requestNext.emit()  # triggers Next GUI thread from event loop
+            except Exception as e:
+                self.valid = False
+                self.error = repr(e)
+                _logger.exception(repr(e))
+            finally:
+                self.busy = False
+
+        t = threading.Thread(
+            target=initialize_device_task,
+            args=(settings, device_id, client),
+            daemon=True)
+        t.start()
+
+    def apply(self):
+        pass
+
+##########################
+#   Import seed wizard   #
+##########################
+
+
+class SatochipSeedLayout(QVBoxLayout):
+    validChanged = pyqtSignal([bool], arguments=['valid'])
+
+    def __init__(self, device):
+        QVBoxLayout.__init__(self)
+
+        label = QLabel(_("Enter a label to name your device:"))
+        self.label_e = QLineEdit()
+        hl = QHBoxLayout()
+        hl.addWidget(label)
+        hl.addWidget(self.label_e)
+        hl.addStretch(1)
+        self.addLayout(hl)
+
+        self.text_e = QTextEdit()
+        self.text_e.setMaximumHeight(60)
+        msg = _("Enter your BIP39 mnemonic:")
+
+        # TODO: validation?
+        def set_enabled():
+            item = ' '.join(str(clean_text(self.text_e)).split())
+            (is_checksum_valid, is_wordlist_valid) = bip39_is_checksum_valid(item)
+            self.validChanged.emit(is_checksum_valid and is_wordlist_valid)
+        self.text_e.textChanged.connect(set_enabled)
```

Comment:
You should really use `SeedLayout` from `electrum/gui/qt/seed_dialog.py` here, and get mnemonic wordlist completion and mnemonic types [electrum,bip39,slip39] entry and validation for free.

Note that BIP39 mnemonics *can* be valid even if words are not in wordlist or checksum failed, see #8720

See https://github.com/spesmilo/electrum/blob/83e14794a1e1202ecfd40f0ea779f0b91ccf2032/electrum/gui/qt/wizard/wallet.py#L597 where this control is used.

Reply:
Thanks for the tip, I will do that and update PR!

Reply:
Well technically, you are using `WCHaveSeed`, not directly `SeedLayout` :)

The options for Electrum and SLIP39 mnemonics are currently enabled for the user, but are not supported further in the setup process. You could generate the BIP32 seed from those mnemonics as well (see `NewWalletWizard.keystore_from_data()`). Note that electrum seeds would skip the derivation path selection.

Alternatively, these mnemonic types should be disabled in the `options` attribute of `SeedLayout`, so the user can't select these.

Also, I think it's useful to add a notice that the passphrase will be integrated to the seed, and not changeable once the user loads the seed onto the card.